### PR TITLE
New data format and producer for PCC objects/event

### PR DESCRIPTION
**Pull request by radla118**
#### PR description:

AlcaPCCProducerEvent  is added with a new PixelClusterCountsPerEvent container to save the PCC objects on an event basis. 

#### PR validation:

Only containers have been added, no sequences in the workflow have changed.

#### if this PR is a backport please specify the original PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by slava77** [comment]
@radla118 
please address https://github.com/cms-sw/cmssw/pull/28807#issuecomment-579179060

tests or further review can not start before the code checks pass

**Comment by christopheralanwest** [comment]
please test

**Comment by christopheralanwest** [comment]
Please give this PR a more meaningful title.

File: DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#include <sstream>
+#include <iostream>
+#include <vector>
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "DataFormats/Luminosity/interface/LumiConstants.h"
+
+namespace reco {
+  class PixelClusterCountsPerEvent {
```
Comment (slava77) [inline DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h]: It's a bit unfortunate that PixelClusterCounts is already used for per lumi data.
"PerEvent" sounds like this has accumulated statistics over many events.
Perhaps "PixelClusterCountsInEvent" is a better choice.

File: DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    void setbxID(unsigned int inputbxID) { bxID = inputbxID; }
+
+    std::vector<int> const& readCounts() const { return (m_counts); }
+
+    std::vector<int> const& readModID() const { return (m_ModID); }
+
+  private:
+    std::vector<int> m_counts;
+    std::vector<int> m_ModID;
+    unsigned int bxID;
```
Comment (slava77) [inline DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h]: please use the same style as the other data members: ```m_bxID```.

There should also be an accessor.

If I were to ignore the style of names in ```PixelClusterCounts```, then the access methods better be called ```counts```, ```modID```, and ```bxID```.

File: DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        m_counts.push_back(0);
+      }
+      m_counts.at(modIndex) += count;
+    }
+
+    void setbxID(unsigned int inputbxID) { bxID = inputbxID; }
+
+    std::vector<int> const& readCounts() const { return (m_counts); }
+
+    std::vector<int> const& readModID() const { return (m_ModID); }
```
Comment (slava77) [inline DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h]: the parentheses are not needed: ```return m_counts;```

File: DataFormats/Luminosity/src/classes_def.xml [hunk 4]
```diff
@@ -49,5 +49,9 @@
   <class name="reco::PixelClusterCounts" ClassVersion="3">
    <version ClassVersion="3" checksum="1474294271"/>
   </class>
- <class name="edm::Wrapper<reco::PixelClusterCounts>"/>
+ <class name="edm::Wrapper<reco::PixelClusterCountsPerEvent>"/>
+  <class name="reco::PixelClusterCountsPerEvent" ClassVersion="1">
+   <version ClassVersion="1"/>
```
Comment (slava77) [inline DataFormats/Luminosity/src/classes_def.xml]: we should start from ```ClassVersion="3"``` https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Class_versioning
The checksum is missing on L54.

File: DataFormats/Luminosity/src/classes_def.xml [hunk 5]
```diff
@@ -49,5 +49,9 @@
   <class name="reco::PixelClusterCounts" ClassVersion="3">
    <version ClassVersion="3" checksum="1474294271"/>
   </class>
- <class name="edm::Wrapper<reco::PixelClusterCounts>"/>
+ <class name="edm::Wrapper<reco::PixelClusterCountsPerEvent>"/>
```
Comment (slava77) [inline DataFormats/Luminosity/src/classes_def.xml]: this should be restored; there is a ```edm::Wrapper<reco::PixelClusterCountsPerEvent>``` on L56 already

File: DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
+#ifndef DataFormats_PixelClusterCountsPerEvent_h
+#define DataFormats_PixelClusterCountsPerEvent_h
```
Comment (slava77) [inline DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h]: Please use an include guard matching the full file name subsystem_package_fileName_h

File: DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+#include <iostream>
+#include <vector>
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "DataFormats/Luminosity/interface/LumiConstants.h"
+
+namespace reco {
+  class PixelClusterCountsPerEvent {
+  public:
```
Comment (slava77) [inline DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h]: there should be a default constructor to initialize the ```bxID``` data member

File: DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    void increment(int mD, int count) {
+      size_t modIndex = std::distance(m_ModID.begin(), std::find(m_ModID.begin(), m_ModID.end(), mD));
+      if (modIndex == m_ModID.size()) {
+        m_ModID.push_back(mD);
+        m_counts.push_back(0);
+      }
+      m_counts.at(modIndex) += count;
```
Comment (slava77) [inline DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h]: the size was already checked; ```m_counts[modIndex]``` can be used instead

File: DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+ * Reconstructed PixelClusterCountsPerEvent object that will contain the moduleID, bxID, and counts per event.
+ *
+ * \authors: Sam Higginbotham (shigginb@cern.ch), Chris Palmer (capalmer@cern.ch), Attila Radl (attila.radl@cern.ch) 
+ * 
+ *
+ */
+#include <algorithm>
+#include <string>
+#include <sstream>
+#include <iostream>
```
Comment (slava77) [inline DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h]: these 3 includes are apparently not needed for this .h file

File: DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+#include <vector>
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "DataFormats/Luminosity/interface/LumiConstants.h"
```
Comment (slava77) [inline DataFormats/Luminosity/interface/PixelClusterCountsPerEvent.h]: these two includes are not needed either.

**Comment by slava77** [comment]
> -code-checks

@radla118 
as previously, 
tests or further review can not start before the code checks pass

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by silviodonato** [comment]
Do you have further comments?
reconstruction: @perrotta @slava77
alca: @franzoni @pohsun @tlampen @tocheng @christopheralanwest

**Comment by pohsun** [comment]
+1

**Comment by slava77** [comment]
> Do you have further comments?

I'll try to get to this soon.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/28807	2654561 
- code changes are in line with the PR description and the follow up review. The reco category part is only in the DataFormats/Luminosity and is not exercised in the standard workflows yet.
- jenkins tests pass and comparisons with the baseline show no differences.

**Comment by silviodonato** [comment]
+1

**Pull request merged**